### PR TITLE
Pickle-less average perceptron tagger

### PR DESCRIPTION
**Pull request by alvations**
Added an implementation to load/save taggers into json format instead of pickle.

Related: #3266 

### Test

```
cd nltk.   # main repo directory
python -m unittest nltk/test/unit/test_pos_tag.py 
python -m doctest nltk/test/tag.doctest  
```

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -7,11 +7,6 @@ repos:
       - id: end-of-file-fixer
       - id: requirements-txt-fixer
       - id: check-yaml
-  - repo: https://github.com/asottile/pyupgrade
-    rev: v3.15.0
-    hooks:
-    - id: pyupgrade
-      args: ["--py38-plus"]
```
Comment (alvations) [inline .pre-commit-config.yaml]: Any clues on why `pyupgrade` is failing on the precommit hooks? I couldn't make it to work even after several different configuration. 

Locally pyupgrade didn't re-write any files. 

Pinging @purificant

**Comment by alvations** [comment]
The CI/CD is failing for some non-deterministic behaviors from ngram LMs unrelated to this PR. 

Anyone has some idea of why that is so for non-py38 tests? Maybe @iliakur would know?

**Review by ekaf — COMMENTED** [review]
@alvations, using 3 json files or only one for each tagger shouldn't make much difference at all.
However, with this PR, the tests fail for me because of a Loading Error, which may be trivial to solve. I will try to see if my setup is somehow wrong.
On the other hand, the same tests succeed using the [one-file approach](https://github.com/nltk/nltk_data/pull/208#issuecomment-2211680961), and a slightly simpler handler. Then, I only see 3 tests failing because of the linebreak discrepancies in the output, which you handle here by reformatting the test output, Alternatively, I think there exists a doctest flag to ignore such trivial differences.

**Comment by alvations** [comment]
For the unittest, you might have to clear your local `nltk_data` cache, usually at `~/nltk_data/tag/`, delete the artifacts inside, and then do a `python -m nltk.downloader average_perceptron_tagger_eng`, before the `python -m unittest ...` 


```
git clone https://github.com/nltk/nltk.git
cd nltk
git checkout  pickle-patch
python -m nltk.downloader average_perceptron_tagger_eng
python -m unittest nltk/test/unit/test_pos_tag.py -vvv
python -m doctest nltk/test/tag.doctest  
```

----

For, the doctest, it should pass with the edits in this PR 

Copy and paste the changes from https://github.com/nltk/nltk/blob/1ec5bacda110624d2e3f310de4f865ccb03b4cd6/nltk/test/tag.doctest to your local machine and retry.

Somehow, I think we haven't used the doctest much in the past and overlooks minor details there. There was even an expected output that was just `[...]` (which it shouldn't) before the fix =) 

Most important line to check in the test for model port integrity should be:

```
    >>> print(tagger.accuracy(gold_data)) # doctest: +ELLIPSIS
    0.885931...
```

**Comment by ekaf** [comment]
My problem was that I had used git to clone nltk_data, so it was not properly set up: there was a _packages_ directory with a _taggers_ subdirectory in it, instead of _taggers_ residing at the top level. Others may have the same problem, because the error messages that I saw reminded much of those reported in several issues related to packages download.

Now all tests pass, including this additional call:
`python -m doctest nltk/tag/perceptron.py -v
`

**Comment by ekaf** [comment]
Concerning the changed linebreaks in doctest output, it may be safer to add the flag:
`# doctest: +NORMALIZE_WHITESPACE`
This  eliminates the need to reformat the output if the linebreaks change in the future.

**Comment by alvations** [comment]
Thanks @ekaf for checking and the suggestion on the doctest option! 

This PR should be good to go once all the CI/CD tests check through, will merge it after.

**Pull request merged**